### PR TITLE
[#187364753] Buyer should be able to checkout

### PR DESCRIPTION
**Pull request by EricNiyo73**
**What does this PR do?**
This PR handle buyer checkout processes, including validation and handling of different payment methods.

**Description of Task to be completed?**

- The user prompts the the delivery informations like address ,delivery date  etc
- Update the Payment Info structure to support both credit card and mobile money payment methods.
- Emit an event on order creation to handle notifications.

**How should this be manually tested?**

Watch video [here](https://www.loom.com/share/e439edaecfee43ef9b0d1b64d8127dab?sid=ac3ddbc0-9b7a-41f8-a1da-d6107ae4e421)
image

#### What are the relevant pivotal tracker/Trello stories?
[Delivers #187364753]

### CODE REVIEW

File: src/validations/orderValidation.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import Joi from 'joi'
+const orderSchema = Joi.object({
+  fullName: Joi.string().required(),
+  phoneNumber: Joi.string()
+    .regex(/^(078|079|072|073)\d{7}$/)
```
Comment (munezerobagira) [inline src/validations/orderValidation.ts]: You use joi object to validate the phone number instead of using your own

**Review by munezerobagira — CHANGES_REQUESTED** [review]

**Review by Bananayosostene — APPROVED** [review]

File: test/middlewares.spec.ts [hunk 2]
```diff
@@ -390,3 +392,79 @@ describe('socketAuthMiddleware', () => {
...
+    expect(next.calledWith()).to.be.true
+  })
+
+  it('should return 404 if no active cart is found', async () => {
+    const findOneStub = sinon.stub(Cart, 'findOne').resolves(null)
+
+    await checkCartMiddleware(req as Request, res as Response, next)
+
+    expect(findOneStub.calledOnce).to.be.true
+    // expect(res.status.calledWith(404)).to.be.true
```
Comment (Bananayosostene) [inline test/middlewares.spec.ts]: remove those comments if they are useless

**Review by Bananayosostene — APPROVED** [review]

**Review by Furaha-Justine — APPROVED** [review]
Good Job

**Review by AngeloChristian1 — APPROVED** [review]

File: src/controllers/checkoutController.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+      return res.status(400).json({ message: 'Invalid payment method' })
+    }
+    try {
+      const cart = await Cart.findOne({
+        where: { buyerId: user.id, status: 'active' },
+      })
+      const totalAmount = cart.items.reduce(
+        (total, item) => item.price * item.quantity,
+        0,
+      )
```
Comment (munezerobagira) [inline src/controllers/checkoutController.ts]:
```suggestion
      const totalAmount = cart.items.reduce(
        (total, item) =>total+ item.price * item.quantity,
        0,
      )
```

**Review by munezerobagira — CHANGES_REQUESTED** [review]

**Review by munezerobagira — APPROVED** [review]

**Pull request merged**